src/main/java/io/jenkins/blueocean/autofavorite/FavoritingScmListener.java

```diff
@@ -39,44 +46,78 @@
-                // Fetch the first commit
-                git.changelog()
-                        .includes(lastBuiltRevision.getSha1())
-                        .to(writer)
-                        .max(1)
-                        .execute();
+        GitClient git = Git.with(TaskListener.NULL, new EnvVars())
+                .in(new File(workspace.getRemote()))
+                .getClient();
 
-                // Parse the changelog
-                GitChangeLogParser parser = new GitChangeLogParser(true);
-                try (StringReader input = new StringReader(writer.toString())) {
-                    List<String> lines = IOUtils.readLines(input);
-                    changeSets = parser.parse(lines);
-                }
+        List<GitChangeSet> changeSets;
+        try (StringWriter writer = new StringWriter()) {
+            // Fetch the first commit
+            git.changelog()
+                    .includes(lastBuiltRevision.getSha1())
+                    .to(writer)
+                    .max(1)
+                    .execute();
+
+            // Parse the changelog
+            GitChangeLogParser parser = new GitChangeLogParser(true);
+            try (StringReader input = new StringReader(writer.toString())) {
+                List<String> lines = IOUtils.readLines(input);
+                changeSets = parser.parse(lines);
             }
+        }
 
-            GitChangeSet first = Iterables.getOnlyElement(changeSets, null);
-            if (first != null) {
-                Job<?, ?> job = build.getParent();
-                User author = first.getAuthor();
-                if (!User.getUnknown().equals(author) && !Favorites.hasFavorite(author, job) && !Favorites.isFavorite(author, job)) {
-                    Favorites.addFavorite(author, job);
-                    logger.log(Level.INFO, "Automatically favorited " + job.getFullName() + " for " + author);
+        GitChangeSet first = Iterables.getOnlyElement(changeSets, null);
+        if (first == null) {
+            return;
+        }
+
+        final Job<?, ?> job = build.getParent();
+        User author = first.getAuthor();
+
+        /**
+         * Here's some fun - depending on the value of {@link GitSCM#isCreateAccountBasedOnEmail) the user returned by
+         * {@link GitChangeSet#getAuthor()} will either created in memory or create on disk (making it a user that can be logged into)
+         * That means if isCreateAccountBasedOnEmail() is true it create any Author it comes across in the Git log as a Jenkins users.
+         * So in that case we can no-op the linear search for a matching email alias.
+         */
+        if (User.getById(author.getId(), false) == null || !((GitSCM) scm).isCreateAccountBasedOnEmail()) {
+            UserProperty property = first.getAuthor().getProperty(UserProperty.class);
+            if (property != null && property.getConfiguredAddress() != null) {
+                for (User user : User.getAll()) {
```

Comment:
Linear search feels wrong here for a few reasons:

1. log(n) time to search here should not be added to the build time of the first run. Probably should offload this to a different thread?
2. I suspect it is expensive to load all Users like this but I cant be sure.

Advice wanted :)